webnn/resources/utils.js

```diff
@@ -1,61 +1,369 @@
 'use strict';
 
-const ExecutionArray = ['async', 'sync'];
+const ExecutionArray = ['sync', 'async'];
 
 // https://webmachinelearning.github.io/webnn/#enumdef-mldevicetype
 const DeviceTypeArray = ['cpu', 'gpu'];
 
+// https://webmachinelearning.github.io/webnn/#enumdef-mloperandtype
+const TypedArrayDict = {
+  float32: Float32Array,
+  int32: Int32Array,
+  uint32: Uint32Array,
+  int8: Int8Array,
+  uint8: Uint8Array,
+};
+
+const sizeOfShape = (array) => {
+  return array.reduce((accumulator, currentValue) => accumulator * currentValue, 1);
+};
+
+/**
+ * Get JSON resources from specified test resources file.
+ * @param {String} file - A test resources file path
+ * @returns {Object} Test resources
+ */
+const loadResources = (file) => {
+  const loadJSON = () => {
+    let xmlhttp = new XMLHttpRequest();
+    xmlhttp.open("GET", file, false);
+    xmlhttp.overrideMimeType("application/json");
+    xmlhttp.send();
+    if (xmlhttp.status == 200 && xmlhttp.readyState == 4) {
+      return xmlhttp.responseText;
+    } else {
+      throw new Error(`Failed to load ${file}`);
+    }
+  };
+
+  const json = loadJSON();
+  return JSON.parse(json.replace(/\\"|"(?:\\"|[^"])*"|(\/\/.*|\/\*[\s\S]*?\*\/)/g, (m, g) => g ? "" : m));
+};
+
+/**
+ * Get exptected data from given resources with output name.
+ * @param {Array} resources - An array of expected resources
+ * @param {String} outputName - An output name
+ * @returns {Number[]} An expected data array
+ */
+const getExpectedData = (resources, outputName) => {
+  let data;
+  for (let subResources of resources) {
+    if (subResources.name === outputName) {
+      data = subResources.data;
+      break;
+    }
+  }
+  if (data === undefined) {
+    throw new Error(`Failed to get expected data sources by ${outputName}`);
+  }
+  return data;
+};
+
+/**
+ * Get ULP tolerance of reduceMean, reduceProduct, reduceSum operations.
+ * @param {Object} resources - Resources used for building a graph
+ * @param {String} operationName - An operation name
+ * @returns {Number} A tolerance number
+ */
+const getReductionPrecisionTolerance = (resources, operationName) => {
+  const inputShape = resources.inputs[Object.keys(resources.inputs)[0]].shape;
+  const rank = inputShape.length;
+  const options = {...resources.options};
+  let sizes;
+  if (options && options.axes) {
+    sizes = options.axes.map(
+                (axis) => axis < 0 ? inputShape[axis + rank] : inputShape[axis]
+    );
+  } else {
+    sizes = inputShape;
+  }
+  let tolerance = sizes.reduce(
+                      (accumulator, currentValue) => accumulator * currentValue
+  );
+  if (['reduceMean'].includes(operationName)) {
```

Comment:
I don't know Javascript well, but is this better than just a string comparison? `operationName == 'reduceMean'` is clearer to me. 🤔

Reply:
Thanks.
I use this condition for extensibility, maybe other reduce operations need adopt this same tolerance later, so we just update array to solve this case, for example
```js
if (['reduceMean', 'reduceXXXX'].includes(operationName)) 
```

Reply:
Ah. Putting the thing being tested after the things being tested against feels reversed :/. Alas, one would think you could readably say...

`if (operationName in ['reduceMean', 'reduceXXXX'])`

...but reading about JS's `in` operator, it behaves counterintuitively, returning true for things that are *not* in the list if an undefined value exists -_-. Oh well.

Reply:
Updated. Thanks.